### PR TITLE
Add support for Webpack 4

### PR DESCRIPTION
**Pull request by phun-ky**
This code is heavily copied from https://github.com/kisenka/svg-sprite-loader/pull/247, but with some adjustments to actually make it work with `webpack@4`. Sorry @rynclark :(

> Good artists copy, great artists steal

This removes support for `webpack@1`, `webpack@2` and `webpack@3`

I've also updated the tests to be run, I couldn't run the tests without babel, since we're only supporting `node@^6.13.0` atm.

### What kind of change does this PR introduce? (bugfix, feature, docs update, improvement)

Adds support for webpack 4

### What is the current behavior? (You can also link to an open issue here)

It doesn't work with webpack 4

### What is the new behavior (if this is a feature change)?

It makes it work with webpack 4

### Does this PR introduce a breaking change?

Yes. Does no longer support `webpack@<4`

### CODE REVIEW

**Comment by rynclark** [comment]
Don't apologise! If you can get the tests to pass you'd be my hero. Gave up last night trying to get it all to work with backwards compatibility.

File: lib/plugin.js [hunk 1]
```diff
@@ -50,106 +46,117 @@ class SVGSpritePlugin {
...
+          .tap(NAMESPACE, () => this.replaceSpriteInModuleSource(compilation));
+
+        compilation.hooks.optimizeExtractedChunks
+          .tap(NAMESPACE, chunks => this.replaceInModuleSource(compilation, chunks));
+
+        if (compilation.hooks.htmlWebpackPluginBeforeHtmlGeneration) {
+          compilation.hooks
+            .htmlWebpackPluginBeforeHtmlGeneration
+            .tapAsync(NAMESPACE, (htmlPluginData, callback) => {
+              htmlPluginData.assets.sprites = generateSprites(this, compilation, symbols);
```
Comment (rynclark) [inline lib/plugin.js]: I do think that the two `htmlWebpackPlugin` hooks don't need to be async though, I was just testing something out.

**Comment by rynclark** [comment]
Personally I think we should drop `webpack` 3 and below support and it be released under a new major version.

**Comment by rynclark** [comment]
If you don't have time to support the project anymore @kisenka we could always see about moving it over to the `webpack-contrib` repo.

@michael-ciniawsky @TheLarkInn

File: lib/plugin.js (hunk 1 above)
Comment (phun-ky) [reply lib/plugin.js]: Have no clue about that tbh, but expect some input from @kisenka anytime soon. (I hope)

**Comment by phun-ky** [comment]
@rynclark I agree on the drop of support for older webpack versions. If other needs it,, they can just lock the versions.

**Comment by klimashkin** [comment]
Dropping webpack < 4 is fine if svg-sprite-loader gets major version bump.
Thanks for PR

File: lib/loader.js [hunk 2]
```diff
@@ -17,26 +16,26 @@ module.exports = function loader(content) {
 
   const done = this.async();
   const loaderContext = this;
-  const { resourcePath, resource, context, loaderIndex } = loaderContext;
+  const { resourcePath, context, loaderIndex } = loaderContext;
```
Comment (michael-ciniawsky) [inline lib/loader.js]: `context` was renamed to `rootContext` in `webpack >= 4` and needs to be checked via

```js
const context = loaderContext.rootContext || loaderContext.context || ...
```

File: lib/loader.js [hunk 3]
```diff
@@ -17,26 +16,26 @@ module.exports = function loader(content) {
...
 
   if (!content.includes('<svg')) {
-    throw new Exceptions.InvalidSvg(content, matchedRules);
+    throw new Exceptions.InvalidSvg(content);
   }
 
+  const configObj = { context: loaderContext };
+
+  configObj.config = getOptions(loaderContext);
+  configObj.target = context.target || loaderContext.target || (context.options && context.options.target);
```
Comment (michael-ciniawsky) [inline lib/loader.js]: `(context.options && context.options.target)` is legacy and could be removed

File: lib/plugin.js [hunk 4]
```diff
@@ -50,106 +46,117 @@ class SVGSpritePlugin {
...
+
     const { symbols } = this.svgCompiler;
 
-    // Handle only main compilation
-    compiler.plugin('this-compilation', (compilation) => {
-      // Share svgCompiler with loader
-      compilation.plugin('normal-module-loader', (loaderContext) => {
-        loaderContext[NAMESPACE] = plugin;
-      });
+    if (useNewTapableAPI) {
```
Comment (michael-ciniawsky) [inline lib/plugin.js]: Better to move the hook callbacks out and feature detect via

```js
const thisCompilationFn = (compilation) => ...

...

if (compiler.hooks) {
  const plugin = { name: 'SVGSpritePlugin' }
   
  const { thisCompilation, ... } = compiler.hooks

  thisCompilation.tap(plugin, thisCompilationFn)
} else {
  compiler.plugin('this-compilation',  thisCompilationFn)
}
```

File: lib/loader.js (hunk 2 above)
Comment (phun-ky) [reply lib/loader.js]: Hm, `this.context` and `this.rootContext` is not the same. `this.context` is the full path to the icons beeing collected, while `this.rootContext` is the path of the project using `svg-sprite-loader`.

Comment (phun-ky) [reply lib/loader.js]: But it seems to me that it doesn't  matter, due to `stringifyRequest` using `path.relative` and node's way to lookup imports, to get the correct import.

File: lib/loader.js (hunk 3 above)
Comment (phun-ky) [reply lib/loader.js]: Removed.

File: lib/plugin.js (hunk 4 above)
Comment (phun-ky) [reply lib/plugin.js]: I'm leaving out that kind of refactoring for now.

**Comment by phun-ky** [comment]
Still need some help with the tests. I think most of the tests failing is due to the plugin and extract part of this loader

**Comment by rynclark** [comment]
I'll take a look later. Mine were failing on the extract part too.

**Comment by kisenka** [comment]
Sorry guys, I was totally busy last few months. So which PR I need to review? :)

**Comment by phun-ky** [comment]
@kisenka this one :) Need help with the tests for extract/plugin part

**Comment by Hypnosphi** [comment]
> This removes support for webpack@1, webpack@2 and webpack@3

Why? I thought loaders can be migrated in a backwards-compatible way

**Comment by phun-ky** [comment]
@Hypnosphi that may be true, but this loader can still be utilized with version locking if `webpack@4` is not an option

**Comment by rynclark** [comment]
I believe I've solved this in #250 whilst keeping backwards compatibility.

**Pull request closed without merging**